### PR TITLE
Start value prioritization

### PR DESCRIPTION
**Pull request by henrikt-ma**
According to [the last web meeting](https://github.com/modelica/ModelicaSpecification/pull/2748#issuecomment-914116630), I'm setting up this PR for filling in some details and documenting why we're ticking off this item on the road map.

Initiating this PR in _Draft_ state, as we need to either pick one one of the designs described in [differences.md](https://github.com/modelica/ModelicaSpecification/blob/MCP/0031%2Bstart-value-prioritization/RationaleMCP/0031/differences.md#guess-value-prioritization) or come up with some other design, and then should also update [syntax.md](https://github.com/modelica/ModelicaSpecification/blob/MCP/0031%2Bstart-value-prioritization/RationaleMCP/0031/syntax.md) before merging this PR.

### CODE REVIEW

**Comment by henrikt-ma** [comment]
To start the discussion, I suggest that we proceed with the design where start value priorities are expressed using the two argument form of `guess`, as in:
```
Real 'x';
parameter equation guess('x', 3) = 1.0; /* Guess value priority is 3 */
```

**Comment by olivleno** [comment]
Web Meeting: Martin, Hans, Gerd, Kai, Henrik, Oliver 
Proposal 1: (from above by Henrik)
> ```
> Real 'x';
> parameter equation guess('x', 3) = 1.0; /* Guess value priority is 3 */
> ```
pro: Allows to attach a priority to any variable, e.g. in an initial equation

Proposal 2: (Hans)
> ```
> Real 'x';
> parameter equation 'x' = guess(1.0, 3); /* Guess value priority is 3 */
> ```
pro: more similar to Modelica and not missing anything

Proposal 3:
> ```
> Real 'x';
> parameter equation guess('x') = prioritized(1.0, 3); /* Guess value priority is 3 */
> ```
pro: more intuitive to read as proposal 2

Discussion:
Hans: We prioritize the value, not the variable.
Henrik: Still one can get the information which value to use.

Use case: Alias elimination
Henrik: The priority can be propagated properly.
In an implementation there is: 
* variable
* guess value
* priority
carried around together.

Use case: start value priority for solving a system of equations

In prop 2 only variable with a parameter equation can get a priority.

Conclusion:
We need some other examples.

**Comment by casella** [comment]
I prepared a suite of examples for OpenModelica [here](https://trac.openmodelica.org/OpenModelica/ticket/5813). For each of them it is clear from an end-user perspective what stat value should be prioritized. However, I could not come up with actual general rules to obtain the required behaviour in all cases. I hope you can.

**Comment by casella** [comment]
```
package Aliases
  type Temperature = Real(start = 300, unit="K");
  type Power = Real(unit="W");

  connector Port
    Temperature T;
    flow Power Q;
  end Port;

  model TemperatureSource
    parameter Temperature Tstart = 320;
    Temperature T1;
    Temperature T2(start = 310);
    Temperature T3(start = Tstart);
    Port port;
  equation
    port.T = T1;
    T1 = T2;
    T2 = T3;
    T1+1e-6*sin(T1) = 300 "Implicit equation on T1";
  end TemperatureSource;
  
  model PowerSource
    parameter Temperature Tstart = 320;
    Port port;
    Temperature T1;
    Temperature T2(start = 330);
    Temperature T3(start = Tstart);
  equation
    port.T = T1;
    T1 = T2;
    T2 = T3;
    port.Q - (port.T-300) = sin(time);
  end PowerSource;

  model WrappedTemperatureSource
    parameter Temperature Tstart = 345;
    Port port;
    TemperatureSource ts(Tstart = Tstart);
  equation
    connect(ts.port, port);
  end WrappedTemperatureSource;  
  
  model WrappedPowerSource
    parameter Temperature Tstart = 345;
    Port port;
    PowerSource ps(Tstart = Tstart);
  equation
    connect(ps.port, port);
  end WrappedPowerSource;  
  
  model System1
    TemperatureSource ts;
    PowerSource ps;
  equation
    connect(ts.port,ps.port);
  end System1;
  
  model System2
    TemperatureSource ts(Tstart = 350);
    PowerSource ps;
  equation
    connect(ts.port,ps.port);
  end System2;
  
  model System3
    TemperatureSource ts(T1(start = 360));
    PowerSource ps;
  equation
    connect(ts.port,ps.port);
  end System3;
  
  model System4
    WrappedTemperatureSource ts(Tstart = 370);
    PowerSource ps;
  equation
    connect(ts.port,ps.port);
  end System4;
  
  model System5
    TemperatureSource ts;
    WrappedPowerSource ps(Tstart = 380);
  equation
    connect(ts.port, ps.port);
  end System5;

  model System6
    WrappedTemperatureSource ts(ts(T1(start = 390)));
    WrappedPowerSource ps(Tstart = 400);
  equation
    connect(ts.port, ps.port);
  end System6;
end Aliases;
```

**Comment by casella** [comment]
At the end of the day, after alias elimination, each of these models has two unknowns: a power flow and a temperature. Everything else are aliases.

As an end-user and library developer, I would like the following start values to be selected for the temperature:

- System1: 320
- System2: 350
- System3: 360
- System4: 370
- System5: 380
- System6: 390 

As far as I understand, the current rules of Section 8.6.2 do not guarantee these choices are taken. In fact, with the current 1.16.0-dev version of OMC, I get:

- System1: 330
- System2: 330
- System3: 330
-  System4: 330
-  System5: 300
- System6: 330 

which is never the value I would expect. I haven't checked with Dymola yet, but I suspect the results will be similar.

I have some loosely formulated requirements, but they need to be formalized into actual rules for the compiler
- type defaults have the lowest priority, they should only be used if there are no other start value specification overriding them
- start values set by modifications when instantiating components should have higher priority than start values set by default binding equations in component declarations
- the priority of start values given by parameters should be determined on the basis of where the actual parameter value is specified
- direct specifications of the start value from the top level of the model to be simulated via nested modifiers should have the precedence over the other start values specified insied the lower-level components
- even in case the parameters are evaluated, the priority information should be computed by the frontend (particularly when evaluating parameters!) and then passed along to the backend, in order to be able to make the correct choices when performing alias elimination

**Comment by henrikt-ma** [comment]
@casella your enthusiasm is highly appreciated.  I'm sorry to say, though, that I don't think this PR or even this MCP is the right place to solve your problems.  What we're discussing in this PR is just a means for attaching priorities to… TBD.  What the MCP does to help solving your problems is thus:
- Describing a general purpose initialization priority to be used in Flat Modelica, so that we (tool vendors and those with interest in portability between tools) can isolate the problem of making tools respect the initialization with the highest priority.
- Giving the language group a platform (that is, Flat Modelica initialization priorities) for describing future changes to full Modelica regarding what to prioritize (corresponding to your list of loosely formulated requirements).

**Comment by henrikt-ma** [comment]
Here's an example that illustrates the discussion at the last [web meeting|https://github.com/modelica/ModelicaSpecification/pull/2997#issuecomment-932093648].

Full Modelica:
```
Real x(final start = 1.0);
```

Corresponding Flat Modelica:
```
  Real 'x';
initial equation
  guess('x', 0) = 1.0; /* Guess value priority of 'x' is 0 */
```

**Comment by HansOlsson** [comment]
Regarding Casella's confusion it might be that it would be better to carry out MCP-0031 work in a separate repo.

> Here's an example that illustrates the discussion at the last [web meeting|https://github.com/[/pull/2997](https://github.com/modelica/ModelicaSpecification/pull/2997)#issuecomment-932093648].
> 
> Full Modelica:
> 
> ```
> Real x(final start = 1.0);
> ```
> 
> Corresponding Flat Modelica:
> 
> ```
>   Real 'x';
> initial equation
>   guess('x', 0) = 1.0; /* Guess value priority of 'x' is 0 */
> ```

To me this just indicates that we are going down the wrong path. We are now introducing guess as something that looks like a general function and thus it seems we are also allowing:
```
initial equation
guess('x', 0) = 2.0+ guess('y',0)*3;
guess('x', 0) + guess('x', 1) = 2.0;
```
This is an added complication - and just removing those possibility requires new semantics.

The goal with FlatModelica should be to have simple semantics, and there are two sensible choices:
* guess-value-and-priority for start-value, and only for start-value as that is what is needed in Modelica. So, something like `'x'(start=guess(1.0, 0));` That would work in Modelica when removing instance-hierarchy. This emphasizes that we are prioritizing start-values (for variables) and nothing else.
* guess-priority for initial equations; something like: `initial equation guess('x'=1.0, 0);` This emphasizes that we are prioritizing initial equations; not variables.

**Comment by henrikt-ma** [comment]
> To me this just indicates that we are going down the wrong path. We are now introducing guess as something that looks like a general function and thus it seems we are also allowing:
> 
> ```
> initial equation
> guess('x', 0) = 2.0+ guess('y',0)*3;
> guess('x', 0) + guess('x', 1) = 2.0;
> ```

This is forbidden according to the rule that guess value priority may only be specified once for any variable.  I agree that it may not be the most elegant design in every aspect, but the point of the syntax is that it is compact and flexible.

For comparison, the following wouldn't work, since it isn't clear what is the guess value parameter to which the priority belongs and what are just normal component references in the equation:
```
initial equation
  guess('x' = 'y', 0);  /* Coming from final x(final start = y) in full Modelica. */
```

**Comment by casella** [comment]
@henrikt-ma, I'm not particularly enthusiastic, in fact I think the current status with start value prioritization in standard Modelica (let alone Flat modelica) is not entirely up to end user's needs (i.e., it is a bit lousy). I read the ticket title and I saw some opportunity to get that improved. Sorry for landing on the wrong page.

I just opened a separate ticket, #3008, on this topic, with up-to-date information.

**Comment by henrikt-ma** [comment]
For a moment, let's consider an alternative design that might look more clean at first glance. Allow an "empty initial equation" to be used for just specifying guess value priority (this is just one of many possible syntaxes for giving the guess value priority separately):
```
  Real 'x';
initial equation
  guess('x', 0);  /* Guess value priority of 'x' is 0 */
  guess('x') = 1.0;
```

Of course, it will still be necessary to detect conflicts such as the following, so what is gained compared to the original design is very little as I see it.  The most important point illustrated by this design alternative is therefore that semantic checks are often hard to avoid, and something one might have to accept in one form or another:
```
  Real 'x';
initial equation
  guess('x', 0); /* Guess value priority of 'x' is 0 */
  guess('x', 1); /* Guess value priority of 'x' is 1 */
  guess('x') = 1.0;
```

**Comment by henrikt-ma** [comment]
> @henrikt-ma, I'm not particularly enthusiastic, in fact I think the current status with start value prioritization in standard Modelica (let alone Flat modelica) is not entirely up to end user's needs (i.e., it is a bit lousy). I read the ticket title and I saw some opportunity to get that improved. Sorry for landing on the wrong page.

It was precisely your interest in trying to improve on lousy semantics that I called enthusiasm.  This enthusiasm is of course a very good thing, and accidentally landing here gave a healthy injection of fresh air to the otherwise very isolated Flat Modelica discussions.

**Comment by HansOlsson** [comment]
> For comparison, the following wouldn't work, since it isn't clear what is the guess value parameter to which the priority belongs and what are just normal component references in the equation:
> 
> ```
> initial equation
>   guess('x' = 'y', 0);  /* Coming from final x(final start = y) in full Modelica. */
> ```

**For clarity that model would actually work.** That is the key point - we have priorities for start-values, and in this case initial equations - and we need to select which of the guesses to satisfy to get a consistently determined initialization problem.

**Comment by henrikt-ma** [comment]
> **For clarity that model would actually work.** That is the key point - we have priorities for start-values, and in this case initial equations - and we need to select which of the guesses to satisfy to get a consistently determined initialization problem.

While this seems doable, it looks more like a future extension of the Flat Modelica initialization model, not something we would need to express the priorities we have in full Modelica today.

**Comment by HansOlsson** [comment]
> > **For clarity that model would actually work.** That is the key point - we have priorities for start-values, and in this case initial equations - and we need to select which of the guesses to satisfy to get a consistently determined initialization problem.
> 
> While this seems doable, it looks more like a future extension of the Flat Modelica initialization model, not something we would need to express the priorities we have in full Modelica today.

I don't see this point at all.

The start-value priority is to prioritize between different guess-values (or initialization equations) - that is the functionality we should preserve.

**Comment by henrikt-ma** [comment]
> The start-value priority is to prioritize between different guess-values (or initialization equations) - that is the functionality we should preserve.

The only kind of priority I see for initial equations in current full Modelica is associated with a particular variable being solved in that equation.  Maybe I'm missing something here?

To me, `guess('x' = 'y', 0)` looks like a prioritized optional initial equation `'x' = 'y'`, equivalent to `'y' = 'x'`. This reminds of an equation one could deduce from `x(final start = y, fixed = true)`, but then the equation wouldn't be optional and the priority wouldn't be of interest.  This is why I think that prioritized optional initial equations look like a possible future extension.

**Comment by HansOlsson** [comment]
> > The start-value priority is to prioritize between different guess-values (or initialization equations) - that is the functionality we should preserve.
> 
> The only kind of priority I see for initial equations in current full Modelica is associated with a particular variable being solved in that equation. Maybe I'm missing something here?

The important priority for start-values is in 
https://specification.modelica.org/maint/3.5/equations.html#recommended-selection-of-start-values

I don't know where you see variables associated with equations with similar priority.

**Comment by henrikt-ma** [comment]
> I don't know where you see variables associated with equations with similar priority.

I don't, and this why I'd like to leave this for a possible future extension.  What we need now is something for the guess value priority of a variable, not the guess value priority of some general initial equation.

**Comment by HansOlsson** [comment]
> > I don't know where you see variables associated with equations with similar priority.
> 
> I don't, and this why I'd like to leave this for a possible future extension. What we need now is something for the guess value priority of a variable, not the guess value priority of some generic initial equation.

No, what we need right now is guess value priority of start-values, not of variables.

The https://specification.modelica.org/maint/3.5/equations.html#recommended-selection-of-start-values _is the priority of start-values_; not variables. That is why the heading says **selection of start values** (we select start-values based on their priority).

The information necessary to do that selection/prioritization is not naturally present in Flat Modelica, but is present in Modelica.
It doesn't depend on the variable - it depends on where the start-value is modified is applied.

**If** we necessarily want to transform start-values into initial equations then we could handle it by having those initial equations "inherit" the priority from the start-value.

**Comment by olivleno** [comment]
Web Meeting: (Martin, Hans, Henrik, Gerd, Kai, Oliver)
Full Modelica:
```
model TemperatureSource
    parameter Temperature Tstart = 320;
    Temperature T1;
    Temperature T2(start = 310);
    Temperature T3(start = Tstart);
    Port port;
  equation
    port.T = T1;
    T1 = T2;
    T2 = T3;
    T1+1e-6*sin(T1) = 300 "Implicit equation on T1";
  end TemperatureSource;
```

Flat Modelica (proposal 1):
```
model TemperatureSource
    parameter Temperature Tstart = 320;
    parameter equation guess(Tstart, 0) = 320;
    Temperature T1;
    Temperature T2(start = 310);
    Temperature T3(start = Tstart);
    Port port;
  equation
    port.T = T1;
    T1 = T2;
    T2 = T3;
    T1+1e-6*sin(T1) = 300 "Implicit equation on T1";
  end TemperatureSource;
```

**Comment by henrikt-ma** [comment]
This is how I imagine that proposal 3 would be extended in the case of a guess value parameter without its own parameter equation:
```
initial equation
  guess('x') = prioritized(1.0, 0); /* Guess value priority of 'x' is 0 */
```

For comparison, repeating proposal 1 from compare https://github.com/modelica/ModelicaSpecification/pull/2997#issuecomment-932812807:
```
initial equation
  guess('x', 0) = 1.0; /* Guess value priority of 'x' is 0 */
```

To simplify the comparison, proposal 1 could also be expressed using `prioritized` (which in the initial design would only be allowed to wrap a guess value parameter):
```
initial equation
  prioritized(guess('x'), 0) = 1.0; /* Guess value priority of 'x' is 0 */
```

My problem with proposal 1 (as imagined above) is that it is something I can only easily make sense of when the initial equation is in a particular form that allows the variable to which the guess value belongs to be directly identifiable:
```
initial equation
  guess(<component-reference>) = prioritized(<expression>, <priority-value>);
```

While I see that this form covers what we need for current Modelica, it appears as an arbitrary and unnecessary restriction.  If we want Flat Modelica to be a powerful lower level language, it would be better to have a syntax that isn't restricted to either guess value parameter equations or a specific form of initial equations.

What proposal 1 gives us is orthogonality between specification of equations and specification of a variable's guess value priority. I'd gladly discuss also other ways of specifying priorities, as long as this kind of orthogonality is achieved.  For example I'd like to repeat one possibility from https://github.com/modelica/ModelicaSpecification/pull/2997#issuecomment-933351208, slightly modified to fit here:
```
initial equation
  prioritized(guess('x'), 0); /* Guess value priority of 'x' is 0 */
  guess('x') = 1.0;
```

**Comment by gkurzbach** [comment]
> I think that this is also causing a bit of confusing in the naming - prioritize is more connected to the first use, and guess more with the second use, so it seems odd to mix them.
 I always thought, the primary use case for priorization is the second case, where the fixed attrribut is usually set to false. In the first case setting fixed to false looks more like a modellicng error.

**Comment by HansOlsson** [comment]
> I always thought, the primary use case for priorization is the second case, where the fixed attrribut is usually set to false. In the first case setting fixed to false looks more like a modelling error.

There are many models in practice where the initialization is underspecified in that way. This is especially an issue in an equation-oriented language as Modelica that supports index-reduction, since the loss of states implied by index-reduction also require a similar reduction in the number of initial conditions.

Just telling users that it is an error is not helpful; that's why there is a mechanism for giving a priority so that user gets the best possible start-values - combined with appropriate diagnostics.

A somewhat similar, but more explicit, idea is to use start-values for parameters without bindings. That allows the user to quickly build a system and even run it - even though those values normally don't make sense - in contrast to the non-fixed start-values that often are correct.

**Comment by HansOlsson** [comment]
> This would not be allowed according to the current proposal, as there can be at most one specification of a variable's guess value priority. In the end, all you have is initial equations where guess value parameters may occur,

But that seems extra work compared to just having restricted initial equations with priority.

> The currently proposed syntax allows the above to be expressed compactly by giving the priorities somewhere – **but only once** – among the initial equations:
> 
> ```
>   sin(guess(x)) = p;
>   guess(y, 2) + guess(y) =  guess(x, 1);
> ```

But the problem is that by breaking the chain between guess(x) and x, instead of between x and the value you introduce additional problems: in the first you case - you have an initial equation with potentially no solution that is used to compute a guess-value that may not be used at all. That's unwanted flexibility in a low-level language.

Putting the priority on the value in restricted equations reduces that issue considerable by itself (it would be an equation that cannot be evaluated, not an equation that cannot be solved) and also makes it clearer that it can be safely avoided (as it is just a value that is ignored).

**Comment by olivleno** [comment]
Web-Meeting (Martin, Hans, Henrik, Gerd, Oliver):
To Do: Specify precedence of lower or higher priority values in the differences.

Proposals
from Hans: Priorities can only be assigned to arrays of simple type and scalars.
from Gerd: Priorities can only be assigned to components with an explicitly mentioned guess value parameter. 
from Henrik: Allow priorities for any component.

Conclusion:
Accept Gerd's proposal.

Update the PR accordingly [Henrik]

**Comment by henrikt-ma** [comment]
Updated as decided above: https://github.com/modelica/ModelicaSpecification/pull/2997#issuecomment-957304550

Update design can be read here:
https://github.com/modelica/ModelicaSpecification/blob/MCP/0031%2Bstart-value-prioritization/RationaleMCP/0031/differences.md#guess-value-prioritization

Requesting reviews from @gkurzbach and @HansOlsson.

File: RationaleMCP/0031/differences.md [hunk 1]
```diff
@@ -670,98 +670,59 @@ initial equation
...
-```
-parameter equation guess('r') = 'R'(guessPriority(1.1, 1), guessPriority(1.3, 0));
+  Real[3] 'x';
+  parameter equation guess('x') = {1.1, 1.2, 1.3};
+initial equation
+  prioritize('x', 2);
 ```
 
-In case all record members have the same priority:
+Multiple specification is an error.  For example (here `R` is a record type):
```
Comment (gkurzbach) [inline RationaleMCP/0031/differences.md:704]: In full Modelica records do not have start attibutes. Do we really want to define guess values and prioritx to record variables/values?

**Review by gkurzbach — CHANGES_REQUESTED** [review]

File: RationaleMCP/0031/differences.md (hunk 1 above)
Comment (henrikt-ma) [reply RationaleMCP/0031/differences.md:704]: The corresponding full Modelica model would be something like this:
```
model RecordMemberStart
  record R
    Real x;
  end R;

  R r(x(start = 1.1));
equation
  der(r.x) = 0.0;
end RecordMemberStart;
```
Is that a problem?

**Review by gkurzbach — APPROVED** [review]

File: RationaleMCP/0031/differences.md (hunk 1 above)
Comment (HansOlsson) [reply RationaleMCP/0031/differences.md:704]: But this does not explain why we want to this. If we have just one member it seems simpler to give it priority, and if there are they are often given separately.

**Comment by henrikt-ma** [comment]
Web meeting (Hans, Gerd, Martin, Henrik):
Split example of multiple specification in two: one about multiple specification, and one showing the use of a record.  Then good to merge. [Henrik]

**Pull request merged**